Server/JokesOnYou.Web.Api/Services/JokesService.cs

```diff
@@ -12,12 +14,58 @@ namespace JokesOnYou.Web.Api.Services
     public class JokesService : IJokesService
     {
         private readonly IJokesRepository _jokesRepo;
+        private readonly IUserRepository _userRepository;
+        private readonly IUnitOfWork _unitOfWork;
         private readonly IMapper _mapper;
 
-        public JokesService(IJokesRepository jokesRepo, IMapper mapper)
+        public JokesService(IJokesRepository jokesRepo, IUserRepository userRepository, IUnitOfWork unitOfWork, IMapper mapper)
         {
             _jokesRepo = jokesRepo;
             _mapper = mapper;
+            _userRepository = userRepository;
+            _unitOfWork = unitOfWork;
+        }
+
+        public async Task<JokeReplyDto> CreateJokeAsync(JokeCreateDto jokeCreateDto)
+        {
+            TrimAndNormalizeJokeCreateDto(jokeCreateDto);
+
+            var isDuplicate = await _jokesRepo.DoesJokeExist(jokeCreateDto);
+            if (isDuplicate)
+            {
+                throw new AppException("Joke Already exists.");
+            }
+
+            var user = await _userRepository.GetUserAsync(jokeCreateDto.UserId);
+            if (user == null)
+            {
+                throw new AppException($"No user with this id:\"{jokeCreateDto.UserId}\" found in the Database.");
+            }
+
+            var joke = _mapper.Map<Joke>(jokeCreateDto);
+            joke.AuthorId = user.Id;
+            await _jokesRepo.CreateJokeAsync(joke);
+
+            var saved = await _unitOfWork.SaveAsync();
+            if (!saved)
+            {
+                if(await _jokesRepo.GetJokeDtoAsync(joke.Id) == null)
```

Comment:
Are you trying to get the joke after the fail saved?

Reply:
yea i find this whole part weird. I have to look into Unit of work more i think.
What would your advice be how to handle this ?

Reply:
The point of the unitofwork is to save multiple changes at once so if one fails they all fail in order to avoid inconsistent data in our DB.

How id handle it is almost the same as you.
1. make an if trying to save changes: If (!await _unitOfWork.SaveAsync())
                                throw new appexception
2. Now we can map the joke to a jokereplydto (Yes i now remembered we dont need to get the item from the db!) Since i noticed you make a joke here before passing down, that joke is a reference type so it will also be updated here with properties such as the id so this is the same as the item in db because EF is tracking it

Reply:
but if another part of the app saves the database in the mean time this will fail ?
while it actually succeeded. is that possible ?